feat: Use system default color mode, but allow user override

ericvolp12 opened the pull request:
I've implemented a fix for https://github.com/bluesky-social/social-app/issues/682 using radio buttons and moved the setting to the Settings page as suggested in the comments below:

![chrome_GrxyX238hI](https://github.com/bluesky-social/social-app/assets/1617325/19402870-0b5a-4763-8560-ab0bf35783b5)
![chrome_MeFfIhmSL2](https://github.com/bluesky-social/social-app/assets/1617325/96f0b5ce-3db4-406c-806c-740bafb29133)

The default is currently `system` so new users will be using their device default themes when launching the app for the first time but after changing it, will have their preferences stored in the shell state.


OLD:
I've implemented a fix for https://github.com/bluesky-social/social-app/issues/682 that doesn't use a radio group but cycles through 3 modes `['light', 'dark', 'system']` and the `system` mode inherits the device theme properly.

Video Demo:
![chrome_2sS6iwAMYw](https://github.com/bluesky-social/social-app/assets/1617325/6593773a-4b9a-4860-81a5-3008ce189fe7)


When system default theme is dark:
![image](https://github.com/bluesky-social/social-app/assets/1617325/2778caab-7548-4490-a860-c8c53c0f60fc)

When system default theme is light:
![image](https://github.com/bluesky-social/social-app/assets/1617325/b8cc95c6-aa81-48b3-b2da-58ea9151996c)

Regardless of the system default theme:
![image](https://github.com/bluesky-social/social-app/assets/1617325/7b04d5a3-3818-4e31-a965-1b431cc940a2)
![image](https://github.com/bluesky-social/social-app/assets/1617325/86ce5d83-a349-4dff-a993-6eae9e405ad8)

Tested in mobile browser as well and it inherits an android device theme properly.

Couldn't get the expo test app to work on my phone yet but I did update the Native components so they should expose the same behavior.

This change doesn't change the `colorScheme` at all, it just updates how the `colorScheme` is set in `ThemeContext.tsx` so that we inherit if the user has chosen `system` as their theme.

We also default in `shell.ts` to use the system theme to conform to user expectations.

ansh commented:
I have identical code to this for this issue. The problem is a UX one. It's hard to express to users that a singular icon button is a 3-way toggle between light/dark/system on mobile. It's more verbose on web so easier to understand for sure.

ericvolp12 commented:
I took another whack at it using a 3-state radio button, I can't test on mobile at the moment though.

![chrome_7JgJpM8pKt](https://github.com/bluesky-social/social-app/assets/1617325/22876157-9b79-42dd-a552-8bbb3336445e)

ericvolp12 commented:
Current mobile web looks like:

![chrome_veKI1pZg6A](https://github.com/bluesky-social/social-app/assets/1617325/599ddbeb-3419-44c7-9c0f-c9b39b4f0415)

notdaniel commented:
> It's hard to express to users that a singular icon button is a 3-way toggle between light/dark/system on mobile.

I think users are used to the word "automatic" in reference to system theme. Paired with the right icon, it would work fine.

So I notice that a moon icon is being used above, even for daytime. If you switched it to a sun icon, then you could use them this way:

- ☀ Light Theme
- ☾ Dark Theme
- ☀/☾ Automatic

I find the text-only buttons to be way less intuitive, personally

pfrazee commented:
I appreciate what you're aiming for here!

Yeah I think the 3-way state is a bit of a challenge. Maybe we move it out of the drawer/right-nav and into the settings?

parion commented:
> Maybe we move it out of the drawer/right-nav and into the settings?

That's what I was thinking too, something ala this in a Display menu. 

![image](https://github.com/bluesky-social/social-app/assets/4733032/67ff0e74-91f9-4296-acba-b0942656cf3d)

snhasani commented:
@ericvolp12 I cocked these codes last night. But in the end I wasn't happy about the result because of the poor UI and UX. So I didn't create a PR for here but you can check my code and maybe reusing some part of it. I refactor the SelectableButton component and create a new one, GroupButton. And some other stuff that's [here](https://github.com/snhasani/social-app/pull/1/files).

I think we can put appearance settings into a modal like a few of current settings. Like twitter iOS dark mode settings:

![telegram-cloud-photo-size-4-5994759638173727624-y](https://github.com/bluesky-social/social-app/assets/23160/b9f38635-9c7c-4bd2-a98f-b4f517199a9b)

P.S: Running iOS or Android is a pain in ass. I couldn't run them after a couple of hours of miserable tries. It is too hard to contribute on something that you can not run it :D

cc @pfrazee @ansh

ericvolp12 commented:
Okay, I've moved the appearance 3-way-toggle to the Settings page and removed it from the other components.

![chrome_GrxyX238hI](https://github.com/bluesky-social/social-app/assets/1617325/f8da4d7f-acda-4a01-ba37-8f234d3b22ab)
![chrome_MeFfIhmSL2](https://github.com/bluesky-social/social-app/assets/1617325/ae387874-73d6-4d16-bc92-ec366f327463)

snhasani commented on src/view/screens/Settings.tsx:
```diff
@@ -282,6 +283,34 @@ export const SettingsScreen = withAuthRequired(
+            Appearance
+          </Text>
+          <View>
+            <View style={[styles.linkCard, pal.view, styles.selectableBtns]}>
```
Here is an opportunity for refactoring and preventing duplicate codes. You can pull out SelectableBtn from `src/view/com/modals/ContentFilteringSettings.tsx` and create a new one component that reused here and there.

You can use my code from [this](https://github.com/snhasani/social-app/commit/7fdf63869b19205b3fa41590a6525491d19709b0) and [this](https://github.com/snhasani/social-app/commit/a818d13f6231485979696bc8f90900490abf7fcb) commit. It's already there.

rezaaa commented:
Good job guys. I hope you change the component design to something better.

pfrazee commented on src/state/models/ui/shell.ts:
```diff
@@ -189,7 +189,7 @@ export interface ComposerOpts {
 
 export class ShellUiModel {
-  darkMode = false
+  colorMode = 'system'
```
I get this concern but I don't think it's confusing enough to change

pfrazee commented:
Awesome, thanks everybody
